frontend/src/pages/projects/screens/detail/storage/ClusterStorageModal.tsx

```diff
@@ -1,58 +1,117 @@
 import useWillNotebooksRestart from '~/pages/projects/notebook/useWillNotebooksRestart';
 import { useIsAreaAvailable, SupportedArea } from '~/concepts/areas';
+import { Table } from '~/components/table';
+import { getNotebookPVCMountPathMap } from '~/pages/projects/notebook/utils';
+import { MOUNT_PATH_PREFIX } from '~/pages/projects/screens/spawner/storage/const';
+import NotebookRestartAlert from '~/pages/projects/components/NotebookRestartAlert';
 import BaseStorageModal from './BaseStorageModal';
-import ExistingConnectedNotebooks from './ExistingConnectedNotebooks';
-import { isPvcUpdateRequired } from './utils';
+import { handleConnectedNotebooks, isPvcUpdateRequired } from './utils';
+import { storageColumns } from './clusterTableColumns';
+import ClusterStorageTableRow from './ClusterStorageTableRow';
 
 type ClusterStorageModalProps = {
   existingPvc?: PersistentVolumeClaimKind;
   onClose: (submit: boolean) => void;
 };
 
 const ClusterStorageModal: React.FC<ClusterStorageModalProps> = ({ existingPvc, onClose }) => {
-  const { currentProject } = React.useContext(ProjectDetailsContext);
+  const {
+    currentProject,
+    notebooks: { data, loaded },
+  } = React.useContext(ProjectDetailsContext);
   const namespace = currentProject.metadata.name;
-  const [removedNotebooks, setRemovedNotebooks] = React.useState<string[]>([]);
-  const [notebookData, setNotebookData] = React.useState<ForNotebookSelection>({
-    name: '',
-    mountPath: { value: '', error: '' },
-  });
+  const allNotebooks = data.map((currentData) => currentData.notebook);
   const { notebooks: connectedNotebooks } = useRelatedNotebooks(
-    ConnectedNotebookContext.EXISTING_PVC,
+    ConnectedNotebookContext.REMOVABLE_PVC,
     existingPvc?.metadata.name,
   );
 
   const workbenchEnabled = useIsAreaAvailable(SupportedArea.WORKBENCHES).status;
 
-  const {
-    notebooks: removableNotebooks,
-    loaded: removableNotebookLoaded,
-    error: removableNotebookError,
-  } = useRelatedNotebooks(ConnectedNotebookContext.REMOVABLE_PVC, existingPvc?.metadata.name);
+  const hasExistingNotebookConnections = connectedNotebooks.length > 0;
+  const [notebookData, setNotebookData] = React.useState<ClusterStorageNotebookSelection[]>([]);
+  const [newRowId, setNewRowId] = React.useState<number>(1);
+
+  React.useEffect(() => {
+    if (hasExistingNotebookConnections) {
+      setNotebookData(
+        connectedNotebooks.map((connectedNotebook) => ({
+          name: connectedNotebook.metadata.name,
+          mountPath: {
+            value: existingPvc
+              ? getNotebookPVCMountPathMap(connectedNotebook)[existingPvc.metadata.name]
+              : '',
+            error: '',
+          },
+          existingPvc: true,
+          isUpdatedValue: false,
+        })),
+      );
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [existingPvc, hasExistingNotebookConnections]);
```

Comment:
This can be refactored to

```
React.useEffect(() => {
  if (hasExistingNotebookConnections && existingPvc) {
    const mappedData = connectedNotebooks.map((connectedNotebook) => ({
      name: connectedNotebook.metadata.name,
      mountPath: {
        value: getNotebookPVCMountPathMap(connectedNotebook)[existingPvc.metadata.name] || '',
        error: '',
      },
      existingPvc: true,
      isUpdatedValue: false,
    }));
    setNotebookData(mappedData);
  }
}, [existingPvc, hasExistingNotebookConnections, connectedNotebooks]);
```